src/test/java/picard/illumina/IlluminaBasecallsToFastqTest.java

```diff
@@ -132,18 +137,100 @@ public void testDualBarcodes() throws Exception {
+            final PrintWriter writer = new PrintWriter(libraryParams);
+            final String header = reader.readLine();
+            writer.println(header + "\tOUTPUT_PREFIX");
+            while (true) {
+                final String line = reader.readLine();
+                if (line == null) {
+                    break;
+                }
+                final String[] fields = line.split("\t");
+                final File outputPrefix = new File(outputDir, StringUtil.join("", Arrays.copyOfRange(fields, 0, concatNColumnFields)));
+                outputPrefixes.add(outputPrefix);
+                writer.println(line + "\t" + outputPrefix);
+            }
+            writer.close();
+            reader.close();
+
+            runPicardCommandLine(new String[]{
+                    "BASECALLS_DIR=" + baseCallsDir,
+                    "LANE=" + lane,
+                    "RUN_BARCODE=HiMom",
+                    "READ_STRUCTURE=" + readStructureString,
+                    "MULTIPLEX_PARAMS=" + libraryParams,
+                    "MACHINE_NAME=machine1",
+                    "FLOWCELL_BARCODE=abcdeACXX",
+                    "USE_NEW_CONVERTER=true",
+                    "NUM_PROCESSORS=2",
+                    "METRICS_FILE=" + new File(outputDir, "test.metrics")
+            });
+
+            final ReadStructure readStructure = new ReadStructure(readStructureString);
+            for (final File outputSam : outputPrefixes) {
+                for (int i = 1; i <= readStructure.templates.length(); ++i) {
+                    final String filename = outputSam.getName() + "." + i + ".fastq";
+                    compareFastqs(testDataDir, outputSam, filename);
+                }
+                for (int i = 1; i <= readStructure.sampleBarcodes.length(); ++i) {
+                    final String filename = outputSam.getName() + ".barcode_" + i + ".fastq";
+                    compareFastqs(testDataDir, outputSam, filename);
+                }
+                for (int i = 1; i <= readStructure.molecularBarcode.length(); ++i) {
+                    final String filename = outputSam.getName() + ".index_" + i + ".fastq";
+                    compareFastqs(testDataDir, outputSam, filename);
+                }
+            }
+        } finally {
+            TestUtil.recursiveDelete(outputDir);
+        }
+    }
+
+    private void compareFastqs(File testDataDir, File outputSam, String filename) {
+        File f1 = new File(outputSam.getParentFile(), filename);
+        File f2 = new File(testDataDir, filename);
+        FastqReader reader1 = new FastqReader(f1);
+        List<FastqRecord> reads = new ArrayList<>();
+        FastqReader reader2 = new FastqReader(f2);
+        for (FastqRecord record : reader1) {
+            reads.add(record);
+        }
+
+        for (FastqRecord record : reader2) {
+            Assert.assertTrue(reads.contains(record));
```

Comment:
This asserts that reader1 contains all of reader2 but not the reverse. Is that sufficient?

I'm guessing there are expected differences in the files that prohibit comparing them directly?